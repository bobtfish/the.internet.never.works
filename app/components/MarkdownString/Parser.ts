const regexMdLinks = /\[([^[]+)\]\((.*)\)/gm

export type MarkdownLink = {
  match: string
  text: string
  url: string
}

const flatten = function (array: any[]) {
  let newArray: any[] = []

  for (const item of array) {
    if (item === undefined) continue
    if (Array.isArray(item)) {
      newArray = newArray.concat(item)
    } else {
      newArray.push(item)
    }
  }

  return newArray
}

export function parseLinks (source: string | any[]): MarkdownLink[] {
  if (!Array.isArray(source)) source = [source]
  return flatten(
    source.map((md: any) => {
      if (typeof md !== 'string') return undefined
      const matches = md.matchAll(regexMdLinks)
      return Array.from(matches).map((value: unknown, _index: number) => {
        const [match, text, url] = value as string[]
        return { match, text, url }
      })
    })
  )
}
