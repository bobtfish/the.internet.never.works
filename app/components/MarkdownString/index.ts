export { MarkdownString } from './MarkdownString'
export { MarkdownParagraph } from './MarkdownParagraph'