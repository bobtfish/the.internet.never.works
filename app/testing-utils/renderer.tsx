import testRenderer, { ReactTestRenderer, ReactTestRendererJSON } from 'react-test-renderer';
import { MantineProvider } from '@mantine/core';

function TestInnerContents({ children }: { children: React.ReactNode }) {
  return <div id="testInnerContents">{children}</div>
}

function create(ui: React.ReactNode) {
  const tr = testRenderer.create(<MantineProvider><TestInnerContents>{ui}</TestInnerContents></MantineProvider>)
  const tree = tr.toJSON() || [];
  const innerTree = ((tree as ReactTestRendererJSON[]).find((node) => node.type === 'div' && node.props.id === 'testInnerContents')||{children:[]}).children || []
  Object.defineProperty(tr, 'toJSON', {
    writable: true,
    value: () => { return innerTree }
  })  
  return tr
}

const act = testRenderer.act

const renderer = { create, act }

export default renderer

export function getInnerElement(r: ReactTestRenderer) {
  return r.root.findByType(TestInnerContents).props.children;
}