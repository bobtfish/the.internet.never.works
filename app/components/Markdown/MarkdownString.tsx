import { Anchor, Text } from '@mantine/core';
import reactStringReplace from 'react-string-replace';
import { parseLinks } from './Parser'
import type { MarkdownStringProps } from './types'

export function MarkdownString({ markdown, styles, anchorProps, anchorTarget, classNames, component = 'span', ...textProps }: MarkdownStringProps) {
    const anchorStyles = styles?.anchor
    anchorTarget ||= 'self'
    const links = parseLinks(markdown)
    return <Text<any> {...textProps} className={classNames?.root} component={component} styles={{root: styles?.root}}>{
        links.reduce((acc, { match, text, url }, i) => {
            return reactStringReplace(acc, match, (_match, j) => (
                <Anchor {...anchorProps} className={classNames?.anchor} target={'_' + anchorTarget} key={i+'-'+j} href={url} styles={{root: anchorStyles}}>{text}</Anchor>
            ))
        }, markdown)
    }</Text>
}