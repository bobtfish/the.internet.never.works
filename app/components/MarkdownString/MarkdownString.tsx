import reactStringReplace from 'react-string-replace';
import { parseLinks } from './Parser'

export function MarkdownString({ markdown }: { markdown: string | React.ReactNode[] }) {
    const links = parseLinks(markdown)
    return <span>{links.reduce((acc, { match, text, url }) => {
        return reactStringReplace(acc, match, (_match, i) => (
            <a key={i} href={url}>{text}</a>
        ))
    }, markdown)}</span>
}