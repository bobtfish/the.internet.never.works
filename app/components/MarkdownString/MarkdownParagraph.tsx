import { splitParagraphs } from './Parser'
import { MarkdownString } from './MarkdownString'

export function MarkdownParagraph({ markdown }: { markdown: string | React.ReactNode[] }) {
    const paras = splitParagraphs(markdown)
    return <div>{paras.map((para) => {
        // eslint-disable-next-line react/jsx-key
        return <p><MarkdownString markdown={para} /></p>
    }, markdown)}</div>
}