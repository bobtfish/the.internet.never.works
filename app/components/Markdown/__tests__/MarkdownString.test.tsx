import renderer from '~/testing-utils/renderer'

import { MarkdownString } from '../MarkdownString'

test('plain text - no markdown', () => {
    const component = renderer.create(
        <MarkdownString markdown="hello world" />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
})

test('with markdown', () => {
    const component = renderer.create(
        <MarkdownString markdown="[a link](http://www.somesite.com/) hello world" />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
})

test('with multiple links', () => {
  const input = "begin text[a link](http://www.somesite.com/) middle filler [another link](http://www.othersite.com/)end text"
  const component = renderer.create(
    <MarkdownString markdown={input} />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})