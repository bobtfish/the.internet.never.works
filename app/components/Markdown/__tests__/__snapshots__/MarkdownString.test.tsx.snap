// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plain text - no markdown 1`] = `
[
  <span
    className="mantine-focus-auto m_b6d8b162 mantine-Text-root"
    style={{}}
  >
    hello world
  </span>,
]
`;

exports[`with markdown 1`] = `
[
  <span
    className="mantine-focus-auto m_b6d8b162 mantine-Text-root"
    style={{}}
  >
    <a
      className="mantine-focus-auto m_849cf0da m_b6d8b162 mantine-Text-root mantine-Anchor-root"
      data-underline="hover"
      href="http://www.somesite.com/"
      style={{}}
      target="_self"
    >
      a link
    </a>
     hello world
  </span>,
]
`;

exports[`with multiple links 1`] = `
[
  <span
    className="mantine-focus-auto m_b6d8b162 mantine-Text-root"
    style={{}}
  >
    begin text
    <a
      className="mantine-focus-auto m_849cf0da m_b6d8b162 mantine-Text-root mantine-Anchor-root"
      data-underline="hover"
      href="http://www.somesite.com/"
      style={{}}
      target="_self"
    >
      a link
    </a>
     middle filler 
    <a
      className="mantine-focus-auto m_849cf0da m_b6d8b162 mantine-Text-root mantine-Anchor-root"
      data-underline="hover"
      href="http://www.othersite.com/"
      style={{}}
      target="_self"
    >
      another link
    </a>
    end text
  </span>,
]
`;
